proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+
+Previous versions of Swift have accepted the `@objc(CustomName) extension` syntax and simply ignored the custom name, so merely adding a custom category name won't break source compatibility.
+
+## Effect on ABI stability
+
+All `@objc` members of an `@implementation extension`—member implementation or otherwise—have the ABI of an `@objc dynamic` member, so turning one into the other is not ABI-breaking. `@implementation extension` classes generate only Objective-C metadata, not Swift metadata, so existing Objective-C subclasses will continue to function as normal.
+
+Because `@implementation` attributes and member implementations are not printed into module interfaces, this proposal has no direct effect on Swift ABI stability.
+
+## Implications on adoption
+
+`@implementation` extensions that implement categories are back-deployable to Swift 5.0 runtimes and later, and many `@implementation` extensions that implement classes are too. However, if a class's ivar layout cannot be computed at compile time, that class will require new runtime support and will not be back-deployable to old platforms.
+
+Affected classes are ones whose stored properties contain a non-frozen enum or struct imported from another module that has library evolution enabled. (This property is transitive—if your stored properties contain a struct in your own module, but that struct has a stored property of an affected type, that also limits back deployment.) In practice, it is usually possible to work around this problem by boxing affected values in a class or existential, at the cost of some overhead.
+
+> **Note**: Some of the required runtime changes are in the Objective-C runtime, so even a snapshot toolchain will not be sufficient to actually run modules with affected classes. However, you can test the diagnostics and code generation by compiling with the experimental feature flag `ObjCImplementationWithResilientStorage`; OS version 99.99 will be treated as high enough to have the necessary runtime support.
+
+## Future directions
+
+### Extending `@objc` capabilities to extend `@objc @implementation` capabilities
+
+`@implementation` cannot create declarations that aren't supported by `@objc`. The most notable limitations include:
+
+* Free functions, global variables, cases of `NS_TYPED_ENUM` typedefs, and other non-member Objective-C declarations.
+* Global declarations imported as members of a type using `NS_SWIFT_NAME`'s import-as-member capabilities.
+* Factory convenience initializers (those implemented as class methods, like `+[NSString stringWithCharacters:length:]`).
+* `__attribute__((objc_direct))` methods and `@property (direct)` properties.
+* Members with nonstandard memory management behavior, even if it is correctly annotated.
+* Members which deviate from the Objective-C error convention in certain subtle ways, such as by having the `NSError**` parameter in the wrong place.
```

Comment:
I think some of these are different-in-kind and deserve mention in the main proposal text. Namely, those members that would appear in an Objective-C `@interface` block but which can't be implemented by `@implementation` deserve to be a bit more front and center (seems like that would be factory inits, direct members, nonstandard memory management, and error convention mismatches).

Reply:
I've added this note to the "Member Implementations" section:

> Some members cannot be implemented in an `@objc @implementation` extension because `@objc` does not support some of the features they use; see "Future Directions" for more specific discussion of this. These members will have to be moved to a separate category and implemented in Objective-C.

And I've also split the list in future directions in two, with the members of `@interface`s listed separately from the global members. (That's basically the same division you suggested.)